hail/python/hail/expr/expressions/typed_expressions.py

```diff
@@ -3136,6 +3136,29 @@ def save(self, uri):
 
         Env.backend().execute(NDArrayWrite(self._ir, hl.str(uri)._ir))
 
+    def _broadcast_to_same_ndim(self, other, init):
+        if isinstance(other, NDArrayExpression):
+            if self.ndim < other.ndim:
+                return self._broadcast(other.ndim, init), other
+            elif self.ndim > other.ndim:
+                return self, other._broadcast(self.ndim, init)
+
+        return self, other
+
+    def _broadcast(self, n_output_dims, init):
+        assert self.ndim < n_output_dims
+
+        # Right-align existing dimensions and start prepending new ones
+        # to the left: e.g. [0, 1] -> [3, 2, 0, 1]
+        # Based off numpy broadcasting with the assumption that everything
+        # can be thought to have an infinite number of 1-length dimensions
+        # prepended
+        old_dims = range(self.ndim)
+        new_dims = range(self.ndim, n_output_dims)
+        idx_mapping = list(reversed(new_dims)) + list(old_dims)
+
+        return init(NDArrayReindex(self._ir, idx_mapping), tndarray(self._type.element_type, n_output_dims))
```

Comment:
What's the purpose of passing in the general init function here? I only see it being used right now with the NDArrayNumericExpression---could we infer what the right expression type is based on the element type and avoid passing in a general init function? Would that be preferable?

Reply:
I can change it to look at the element type